hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,171 @@
+package com.hazelcast.spi.impl;
+
+import com.hazelcast.instance.Node;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.util.ConcurrencyUtil;
+import com.hazelcast.util.ConstructorFunction;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Monitors the operation threads to find and collect data of long running operations.
+ */
+final class SlowOperationDetector extends Thread {
+
+    private static final long ONE_SECOND_IN_NANOS = TimeUnit.SECONDS.toNanos(1);
+
+    private final BasicOperationScheduler.OperationThread[] genericOperationThreads;
+    private final BasicOperationScheduler.OperationThread[] partitionOperationThreads;
+
+    private final CurrentOperationData[] genericCurrentOperationData;
+    private final CurrentOperationData[] partitionCurrentOperationData;
+
+    private final long slowOperationThresholdNanos;
+    private final long logPurgeIntervalNanos;
+    private final long logLifetimeNanos;
+
+    private final ILogger logger;
+
+    private final ConcurrentHashMap<Integer, SlowOperationLog> slowOperationLogs = new ConcurrentHashMap<Integer, SlowOperationLog>();
+
+    private long nowNanos;
+    private long nowMillis;
+
+    private volatile boolean running = true;
```

Comment:
I think this is bad code style, since you create a negation in the main usage of the variable): "if not state do something"

In this particular case it may be quite easy to parse, but I've seen this approach with very nasty variable names, which gave you a very hard way to understand the meaning of code. Sometimes the variables started with "not", so you had "if not notState do something".

So I got really used to avoid negation of boolean variables in the main code path.